lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -281,28 +287,53 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
     if (shouldCaptureOutput) {
       // Close the write end of the output pipe.
       ::close(outputPipe[1]);
-
-      // Read all the data from the output pipe.
-      while (true) {
-        char buf[4096];
-        ssize_t numBytes = read(outputPipe[0], buf, sizeof(buf));
-        if (numBytes < 0) {
+      ::close(stderrPipe[1]);
+
+      fd_set read_flags;
+      struct timeval waitd = {10, 0};
+      FD_ZERO(&read_flags);
+      FD_SET(stderrPipe[0], &read_flags);
+      FD_SET(outputPipe[0], &read_flags);
+      bool readStdout = false, readStderr = false;
+      while(true) {
+        int select_result = select(stderrPipe[0] + 1, &read_flags, nullptr, nullptr, &waitd);
+        if (select_result < 0) {
           getDelegate().commandProcessHadError(
               context.job.getForCommand(), handle,
-              Twine("unable to read process output (") + strerror(errno) + ")");
+              Twine("unable to use select() while reading from subprocess (") + strerror(errno) + ")");
           break;
         }
+        if (select_result == 0) break;
```

Comment:
This will no longer be the right behavior once the timeout disappears, the result should never be 0 (and should be asserted, probably).